Add Composite aggregation page. (#9479)
opensearch-project/documentation-website: _aggregations/bucket/composite.md

          "avg_spend": {
            "value": 72.1588623046875
          }
        },
        {
          "key": {
            "weekday": "Monday",
            "gender": "MALE"
          },
          "doc_count": 259,
          "avg_spend": {
            "value": 86.1754946911197
          }
        },
        {
          "key": {
            "weekday": "Saturday",
            "gender": "FEMALE"
          },
          "doc_count": 365,
          "avg_spend": {
            "value": 73.53236301369863
          }
        },
        {
          "key": {
            "weekday": "Saturday",
            "gender": "MALE"
          },
          "doc_count": 371,
          "avg_spend": {
            "value": 72.78092360175202
          }
        }
      ]
    }
  }
}
```

## Paginating composite results

If a request results in more than `size` buckets, then `size` buckets are returned. In this case, the result contains an `after_key` object containing the key of the next bucket in the list. To retrieve the next `size` buckets of the request, make the request again, supplying the `after_key` in the `after` parameter. For an example, see the request in [Histogram](#histogram).

> Always use the `after_key` rather than copying the last bucket to continue a paginated response. The two are sometimes not the same.

Reviewer: ```suggestion
Always use the `after_key` rather than copying the last bucket to continue a paginated response. The two are sometimes different.
```

{: .important}

## Improving performance with index sorting

> To speed up composite aggregations on large datasets, you can sort your index using the same fields and order as your aggregation sources. When the `index.sort.field` and `index.sort.order` match the source fields and their order in the composite aggregation, OpenSearch can return results more efficiently and with less memory usage. While index sorting adds minor overhead during indexing, the query performance gains for composite aggregations are significant.

Reviewer: ```suggestion
To speed up composite aggregations on large datasets, you can sort your index using the same fields and order used in your aggregation sources. When the `index.sort.field` and `index.sort.order` match the source fields and order used in the composite aggregation, OpenSearch can return results more efficiently and with less memory usage. While index sorting adds minor overhead during indexing, the query performance gains for composite aggregations are significant.
```

> The following example request sets sort fields and a sort order for each of the fields on the `my-sorted-index` index:

Reviewer: ```suggestion
The following example request sets sort fields and a sort order for each of the fields in the `my-sorted-index` index:
```

```json
PUT /my-sorted-index
{
  "settings": {
    "index": {
      "sort.field": ["customer_id", "timestamp"],
      "sort.order": ["asc", "desc"]
    }
  },
  "mappings": {
    "properties": {
      "customer_id": {
        "type": "keyword"
      },
      "timestamp": {
        "type": "date"
      },
      "price": {
        "type": "double"
      }
    }
  }
}
```
{% include copy-curl.html %}

> The following request creates a composite aggregation on the `my-sorted-index` index. Because the index is sorted on `customer_id` ascending and `timestamp` descending, and the aggregation sources match that sort order, this query runs faster and with reduced memory pressure:

Reviewer: ```suggestion
The following request creates a composite aggregation on the `my-sorted-index` index. Because the index is sorted by `customer_id` in ascending order and `timestamp` in descending order and the aggregation sources match that sort order, this query runs faster and with reduced memory pressure:
```

---
layout: default
title: Composite
parent: Bucket aggregations
nav_order: 17
redirect_from:
  - /query-dsl/aggregations/bucket/composite/
---

> # Composite aggregations

Author: There does not seem to be any mention of index sorting in OpenSearch. Is it possible to speed up `composite` searches on large datasets by sorting an index? If so, is this a common enough use case to mention on this page?

Reviewer 2: This probably warrants a section of its own. I'll delegate you to decide the section title and position in the documentation, but here is a small explanation which I drafted which you can reword and use.

-----

To improve performance on large datasets, you should sort your index to match your composite aggregation query. When the fields defined in the aggregation sources are identical in sequence and sort order to the fields defined in the `index.sort.field` and `index.sort.order` setting, OpenSearch can return results much faster and with significantly less memory pressure. While this adds a minor overhead to indexing, the gains in query performance are substantial for this use case.

For example,

```
PUT /my-sorted-index
{
  "settings": {
    "index": {
      "sort.field": [ "customer_id", "timestamp" ],
      "sort.order": [ "asc", "desc" ]
    }
  },
  "mappings": {
    "properties": {
      "customer_id": {
        "type": "keyword"
      },
      "timestamp": {
        "type": "date"
      },
      "price": {
        "type": "double"
      }
    }
  }
}
```

```
GET /my-sorted-index/_search
{
  "size": 0,
  "aggs": {
    "my_buckets": {
      "composite": {
        "size": 1000,
        "sources": [
          { "customer": { "terms": { "field": "customer_id", "order": "asc" } } },
          { "time": { "date_histogram": { "field": "timestamp", "calendar_interval": "1d", "order": "desc" } } }
        ]
      }
    }
  }
}
```

Because the sources (`customer_id` ascending, `timestamp` descending) match the index sort order, this query will be highly optimized.

          { "unit_price": { "histogram": { "field": "products.base_unit_price", "interval": 50 }}}
        ],
        "size": 6,
        "after": {
            "quantity": 2,
            "unit_price": 150
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

As shown in the following example response, there are only two buckets left:

```json
{
  "took": 12,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 4675,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "composite_buckets": {
      "after_key": {
        "quantity": 2,
        "unit_price": 500
      },
      "buckets": [
        {
          "key": {
            "quantity": 2,
            "unit_price": 200
          },
          "doc_count": 8
        },
        {
          "key": {
            "quantity": 2,
            "unit_price": 500
          },
          "doc_count": 4
        }
      ]
    }
  }
}
```

## Date histogram

Use the `date_histogram` to create composite aggregations of date ranges.

> OpenSearch represents dates, including `date_interval` bucket keys, as `long` integers representing [milliseconds since epoch](https://en.wikipedia.org/wiki/Unix_time). You can output a date format instead using the `format` parameter. This does not change the key order.

Author: Please confirm that the key order does not change. Key order is based on the INT64 value, correct?

OpenSearch stores date-times in UTC. You can display output results in a different timezone using the `time_zone` parameter.

> ### Parameters: Date histogram

Reviewer 2: This is probably not the best place to add details about Date histogram parameters.
Please consider moving the parameters in details to https://docs.opensearch.org/docs/latest/aggregations/bucket/date-histogram/

Similarly for geotile grid: https://docs.opensearch.org/docs/latest/aggregations/bucket/geotile-grid/

The corresponding documentation should have details explicit about parameters. 
Mentions of these aggregations in composite terms should be just to introduce an example and then point out to the detailed documentation on how those can be customized.

  "hits": {
    "total": {
      "value": 4675,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "composite_buckets": {
      "after_key": {
        "product_creation_date": "2016",
        "order_date": "2025-02-23"
      },
      "buckets": [
        {
          "key": {
            "product_creation_date": "2016",
            "order_date": "2025-02-20"
          },
          "doc_count": 146
        },
        {
          "key": {
            "product_creation_date": "2016",
            "order_date": "2025-02-21"
          },
          "doc_count": 153
        },
        {
          "key": {
            "product_creation_date": "2016",
            "order_date": "2025-02-22"
          },
          "doc_count": 143
        },
        {
          "key": {
            "product_creation_date": "2016",
            "order_date": "2025-02-23"
          },
          "doc_count": 140
        }
      ]
    }
  }
}
```

## Geotile grid

Use `geotile_grid` sources to aggregate `geo_point` values into buckets representing map tiles. As with the other composite aggregation `sources`, by default results include only buckets that contain data.

> Each cell corresponds to a [map tile](https://en.wikipedia.org/wiki/Tiled_web_map). Cell labels use a "{zoom}/{x}/{y}" format.

Author: Is there a site that does a good job of explaining the x,y system that geo-grid uses? 

Does OpenSearch use one of the available geo-grid standards?

### Parameters: Geotile grid

> | Parameter      | Required/Optional  | Data type             | Description |

Reviewer 3: Geotile grid page already has a parameter table. Removing this table.

            "value": 71.7733395989975
          }
        },
        {
          "key": {
            "weekday": "Friday",
            "gender": "MALE"
          },
          "doc_count": 371,
          "avg_spend": {
            "value": 79.72514108827494
          }
        },
        {
          "key": {
            "weekday": "Monday",
            "gender": "FEMALE"
          },
          "doc_count": 320,
          "avg_spend": {
            "value": 72.1588623046875
          }
        },
        {
          "key": {
            "weekday": "Monday",
            "gender": "MALE"
          },
          "doc_count": 259,
          "avg_spend": {
            "value": 86.1754946911197
          }
        },
        {
          "key": {
            "weekday": "Saturday",
            "gender": "FEMALE"
          },
          "doc_count": 365,
          "avg_spend": {
            "value": 73.53236301369863
          }
        },
        {
          "key": {
            "weekday": "Saturday",
            "gender": "MALE"
          },
          "doc_count": 371,
          "avg_spend": {
            "value": 72.78092360175202
          }
        }
      ]
    }
  }
}
```

> ## Paginating composite results

Author: How long are results cached for display pagination? 

Are there any shortcuts to displaying result pages, or do you have to give it the entire request plus `after` parameter each time?
